data/tasks.js

```diff
@@ -0,0 +1,192 @@
+import { ObjectId } from "mongodb";
+import utils from "../utils/utils.js";
+import { usersCollection } from "../config/mongoCollections.js";
+
+const tasksDataFunctions = {
+  async getTaskById(id) {
+    utils.checkObjectIdString(id);
+    id = id.trim();
+    const tasks = await tasksCollection();
+    const task = await tasks.findOne({ _id: new ObjectId(id) });
+
+    if (!task) {
+      throw new Error(`Task with ID ${id} not found`);
+    }
+
+    return task;
+  },
+  /**
+   * @param {string} userId - The user ID to which the task is associated
+   * @param {string} title - The title of the task
+   * @param {string} textBody - The body text of the task
+   * @param {Date} dateAddedTo - The date when the task was added to the collection
+   * @param {Date} dateDueOn - The date when the task is due
+   * @param {number} priority - The priority of the task (1, 2, or 3)
+   * @param {string} tag - The custom tag for the task
+   */
+  async createTask(
+    userId,
+    title,
+    textBody,
+    dateAddedTo,
+    dateDueOn,
+    priority,
+    tag
+  ) {
+    utils.checkObjectIdString(userId);
+    utils.validateStringInput(title, "title");
+    utils.validateStringInput(textBody, "textBody");
+    utils.validateDate(dateAddedTo, "dateAddedTo");
+    utils.validateDate(dateDueOn, "dateDueOn");
+    utils.validatePriority(priority);
+    utils.validateStringInput(tag, "tag");
+
+    title = title.trim();
+    dateAddedTo = dateAddedTo.trim();
+    dateDueOn = dateDueOn.trim();
+    textBody = textBody.trim();
+    tag = tag.trim().toLowerCase();
+
+    const users = await usersCollection();
+    const user = await users.findOne({ _id: new ObjectId(userId) });
+    if (!user) {
+      throw new Error("User not found.");
+    }
+    const newTask = {
+      title: title,
+      textBody: textBody,
+      dateCreated: new Date(),
+      dateAddedTo: new Date(dateAddedTo),
+      dateDueOn: new Date(dateDueOn),
+      priority: priority,
+      tag: tag,
+      checked: false,
+      type: "task",
+    };
+
+    const tasks = await tasksCollection();
+    const taskExists = await tasks.findOne({ title: title });
+
+    if (taskExists) {
+      throw new Error(
+        `Task title already exists for the User ${user.first_name}`
+      );
+    }
+    const insertInfo = await tasks.insertOne(newTask);
+
+    if (insertInfo.insertedCount === 0) {
+      throw new Error("Could not add task");
+    }
+
+    const newId = insertInfo.insertedId;
+    await users.updateOne(
+      { _id: new ObjectId(userId) },
+      { $push: { taskIds: newId } }
+    );
+    return await this.getTaskById(newId.toString());
+  },
+
+  async updateTask(id, updatedTask) {
+    utils.checkObjectIdString(id);
+
+    const tasks = await tasksCollection();
+
+    const updatedTaskData = {};
+    updatedTask.title = updatedTask.title.trim();
+    updatedTask.dateAddedTo = updatedTask.dateAddedTo;
+    updatedTask.dateDueOn = updatedTask.dateDueOn;
```

Comment:
In Create task you are persisting the db with Date() object while in update you are purging the db with string. Please check and revert.